### ISSUE TITLE
GADT and Lenses

### ISSUE THREAD

**Issue by elucash** [immutables/immutables#47]
Remarkably, _Immutables_ have pretty decent support for the algebraic data type -like classes.
Examples:
- ADT-like types using OO inheritance and object graphs:
  https://github.com/immutables/immutables/blob/master/generator-processor/src/org/immutables/generator/processor/Trees.java
- immutable object graph transformers: https://github.com/immutables/immutables/blob/master/generator-processor/src/org/immutables/generator/processor/TypeResolver.java

Things that could be improved
1. Visitors. A well-thought-out visitor generation would be a good asset along with tree-transformers which are already implemented.
2. Documentation and examples.

**Comment by elucash** [comment]
Musing about https://gist.github.com/elucash/8542f28198ca571dd666

**Comment by bgertzfield** [comment]
I was JUST going to suggest something like that gist. We use pattern matching in a bunch of languages and it's amazing—way better than chains of `Optional<T>`.

Instead of inheritance, I would suggest an explicit enumerated algebraic type using "either A or B or C" style composition.

**Comment by elucash** [comment]
I still cannot imagine how declaration will look like.
Ok, now starting to imagine something, here's first cut
```java
@Case.Chain(cases = { A.class, B.class, C.class })
interface SomeType {
   @Value.Immutable interface A {}
   @Value.Immutable interface B {}
   @Value.Immutable interface C {}
}
...
// generated class
Object someCase = ...;
SomeTypeCases.match(someCase) // 'match' could be static imported etc
   .caseA(a -> ...)
   .caseB(b -> ...)
   .caseC(c -> ...)
   .result();

// where @Case.Chain
@interface Chain {
   /** Common supertype for all cases. */
   Class<?> supertype() default Object.class;
   /** Case classes, when unspecified, tries to find cases in nested scope */
   Class<?> cases() default {};
}
```

**Comment by io7m** [comment]
I've documented a very common encoding for case analysis of types in Java:

http://io7m.com/documents/java/tt3-genvis/

For the following Haskell type: 

```
data Scalar =
     Float Integer
   | IntegerSigned Integer 
   | IntegerUnsigned Integer
   | IntegerSignedNormalized Integer
   | IntegerUnsignedNormalized Integer
```

Here's an `immutables` encoding of the type:

```
@Value.Enclosing public interface ScalarTypes
{
  /**
   * The type of scalar types.
   */

  interface ScalarType
  {
    /**
     * Accept a matcher.
     *
     * @param m   The matcher
     * @param <A> The type of returned values
     * @param <E> The type of raised exceptions
     *
     * @return The value returned by {@code m}
     *
     * @throws E If {@code m} raises {@code E}
     */

    <A, E extends Exception> A matchScalar(MatcherType<A, E> m)
      throws E;
  }

  /**
   * The type of scalar type matchers.
   *
   * @param <A> The type of returned values
   * @param <E> The type of raised exceptions
   */

  interface MatcherType<A, E extends Exception>
  {
    /**
     * Match a float type.
     *
     * @param t The type
     *
     * @return A value of {@code A}
     *
     * @throws E If required
     */

    A onFloat(FloatType t)
      throws E;

    /**
     * Match a signed integer type.
     *
     * @param t The type
     *
     * @return A value of {@code A}
     *
     * @throws E If required
     */

    A onIntegerSigned(IntegerSignedType t)
      throws E;

    /**
     * Match an unsigned integer type.
     *
     * @param t The type
     *
     * @return A value of {@code A}
     *
     * @throws E If required
     */

    A onIntegerUnsigned(IntegerUnsignedType t)
      throws E;

    /**
     * Match a signed normalized integer type.
     *
     * @param t The type
     *
     * @return A value of {@code A}
     *
     * @throws E If required
     */

    A onIntegerSignedNormalized(IntegerSignedNormalizedType t)
      throws E;

    /**
     * Match an unsigned normalized integer type.
     *
     * @param t The type
     *
     * @return A value of {@code A}
     *
     * @throws E If required
     */

    A onIntegerUnsignedNormalized(IntegerUnsignedNormalizedType t)
      throws E;
  }

  @Value.Immutable(builder = false) abstract class FloatType
    implements ScalarType
  {
    @Value.Parameter abstract int size();

    @Override public final <A, E extends Exception> A matchScalar(
      final MatcherType<A, E> m)
      throws E
    {
      return m.onFloat(this);
    }
  }

  @Value.Immutable(builder = false) abstract class IntegerSignedType
    implements ScalarType
  {
    @Value.Parameter abstract int size();

    @Override public final <A, E extends Exception> A matchScalar(
      final MatcherType<A, E> m)
      throws E
    {
      return m.onIntegerSigned(this);
    }
  }

  @Value.Immutable(builder = false) abstract class IntegerUnsignedType
    implements ScalarType
  {
    @Value.Parameter abstract int size();

    @Override public final <A, E extends Exception> A matchScalar(
      final MatcherType<A, E> m)
      throws E
    {
      return m.onIntegerUnsigned(this);
    }
  }

  @Value.Immutable(builder = false) abstract class IntegerSignedNormalizedType
    implements ScalarType
  {
    @Value.Parameter abstract int size();

    @Override public final <A, E extends Exception> A matchScalar(
      final MatcherType<A, E> m)
      throws E
    {
      return m.onIntegerSignedNormalized(this);
    }
  }

  @Value.Immutable(builder = false) abstract class IntegerUnsignedNormalizedType
    implements ScalarType
  {
    @Value.Parameter abstract int size();

    @Override public final <A, E extends Exception> A matchScalar(
      final MatcherType<A, E> m)
      throws E
    {
      return m.onIntegerUnsignedNormalized(this);
    }
  }
}
```

The advantage of this encoding is that to pattern match on a value of type `ScalarType`, you're forced to create an instance of type `MatcherType` and this instance will become incomplete when a new case is added to the `ScalarType`; the compiler tells you all of the places where your code is now incomplete.

Clearly there's a lot of boilerplate here that could be removed by a generator.

**Comment by io7m** [comment]
Another encoding is to generate a method `match` that has one parameter per possible type case. As an example, a method with the following signature would be
generated on the common supertype of the cases:

```
public abstract <A, E extends Exception> A match(
  PartialFunction<Float, A, E> on_float,
  PartialFunction<IntegerSigned, A, E> on_integer_signed,
  PartialFunction<IntegerUnsigned, A, E> on_integer_unsigned,
  PartialFunction<IntegerSignedNormalized, A, E> on_integer_signed_normalized,
  PartialFunction<IntegerUnsignedNormalized, A, E> on_integer_unsigned_normalized);
```
... and the implementation for the `IntegerSigned` type, for example, would be:

```
@Override public <A, E extends Exception> A match(
  PartialFunction<Float, A, E> on_float,
  PartialFunction<IntegerSigned, A, E> on_integer_signed,
  PartialFunction<IntegerUnsigned, A, E> on_integer_unsigned,
  PartialFunction<IntegerSignedNormalized, A, E> on_integer_signed_normalized,
  PartialFunction<IntegerUnsignedNormalized, A, E> on_integer_unsigned_normalized)
{
  return on_integer_signed.apply(this);
}
```

This has the questionable advantage that at least for Java 8, the VM is likely to do a much better job of avoiding allocations for the passed in functional interfaces.

I think the critical requirements are that pattern matching be as cheap as possible (because once you start using it, you start using it _everywhere_), and that when new cases are added to a particular algebraic type, the encoding of the type is such that all pattern matches that don't mention the new case now yield compilation errors. When I should be handling a new case, I want the compiler to tell me about it!

**Comment by io7m** [comment]
The latter encoding is something that I would not want to write by hand, because it's very easy to make a mistake. It would however be extremely safe if that code was generated!

**Comment by elucash** [comment]
@io7m  Interesting! The latter is called _catamorphism_ iirc.
You're probably already know these, but I'll just list some other projects which explore AGDT encoding:

* https://github.com/derive4j/derive4j
* https://github.com/sviperll/adt4j
* https://github.com/talios/javagadt
* I'm sure there are others...

While there are plans to include something like this in Immutables and want it be really great, we are currently struggle with fundamental things like adequate support for generics. As soon as we get closer I hope you might help to sort out structure/apis etc. If you want to experiment, you may try to hack something using [Transformers.generator](https://github.com/immutables/immutables/blob/master/value-processor/src/org/immutables/value/processor/Transformers.generator) template as a starting point. There's an example of some limited support for subtypes. (annotations are placed in experimental `cases` module, but implementations are in `value-processor` module, along with other generator implementations).

**Comment by io7m** [comment]
Hm!

I'm familiar with both `derive4j` and `adt4j` and have had issues with both. That's the first time I've ever seen that `javagadt` package though, it looks significantly better than the others.

I'll try to find some time to assist.

**Comment by io7m** [comment]
Wrote this article recently that may be relevant:

http://io7m.com/documents/adt-jvm/

**Comment by elucash** [comment]
Nice read! This should be like a handbook to read before inventing/implementing generator for the GADT encoding.

**Comment by before** [comment]
Thank you @io7m for sharing these insides with us. I enjoyed reading it.

@elucash are you considering to generate code like that? I think a lot of us are interested into it.

**Comment by talios** [comment]
@io7m thanks for including my `javagadt` project in that document - one of the things you mentioned was the overloaded syntax I used - which I'd based of how https://github.com/JamesIry/jADT implemented it.  I'm actually tempted to replace my usage of jparsec for antlr4 ( internal change ) and change the code that's generated to take the expanded arguments.

One other idea I had floating around was to drop the Maven plugin in favour of an annotation processor, but keep the `.gadt` file, and having a `package` only `@Gadt` annotation which you put in `package-info.java` and look for an associated `package.gadt` file.

But this isn't really the place to discuss that...

**Comment by elucash** [comment]
@before There are still plans to do it. It's likely it will be a new java8 only codebase, (quite tired of java6/7/8 interoperability issues and older eclipse/javac compiler issues, maybe it will be Immutables 3). Frankly, I still not decided if to continue the java model approach, the best what we could get is the syntax like this:

```java
// sum type
@Data
public interface List<T> extends DataList.List {
  static <T> List<T> cons(T head, List<T> tail) {
    return DataList.List.cons(head, tail);
  }
  static <T> List<T> nil() {
    return DataList.List.nil();
  }
}

// or regular record with builder
@Data
public interface Record extends DataRecord.Record {
  long id();
  String name();
  class Record extends DataRecord.Record.Builder {}
}
```

I try to maintain that only the user declared type is used elsewhere and utility methods will be defined by using special "sandwich" generation strategy where user declared type both extends a generated type and is being extended by a generated type.

Having that said still don't know when will be ready dive into this

@talios Annotation processor might be a way to go as you get the integration with any build system which should support standard annotation processors (in theory). The approach with a package-info/custom syntax file worked fine for me as I used it for some generators. Works even in IDE, but have to either make recompile/clean or change/touch package file to trigger regeneration on changes.

**Comment by io7m** [comment]
I've been considering approaching the OpenJDK people for a while with some sort of proposal/request to extend Java with real case analysis. As far as I can tell, it would require a few non-intrusive changes and would be far better than a library implementation of this stuff.

* Add syntax in Java to declare an algebraic type. I think the least intrusive language change that could accomodate this would imitate the approach already taken by other languages: Mark a class as "sealed" in some manner, and then assume that all inner subclasses are cases of the resulting algebraic type. The `switch` keyword could be reused here:

```
switch class T
{
  class A extends T { }
  class B extends U { }
  class C extends V { }
}
```

The implementation should also support nested `switch` classes (which isn't any more difficult than supporting them *at all*).

* Extend the syntax of `switch` statements to accept values of `switch` classes, with each branch representing a constructor case. They obviously already extended the `switch` statement with enums in Java 5, and again with strings in Java 7, so it's unlikely to be too controversial a change for the generally conservative designers. The important thing is to ensure that the compiler does exhaustiveness checks, which it does do in some manner for `enum` types already. Overlapping cases are already rejected by the existing `switch` statement. Borrowing from the lambda syntax a little:

```
switch (t) {
  (A a): { 
    // a == t
    break;
  }
  (B b): {
    // b == t
    break;
  }
  (C c): {
    // c == t
    break;
  }
}
```

The compiler would implicitly ensure that, for example, in the `A` case, `t instanceof A` and would expose an alias of `t` called `a` that had been implicitly casted to `A`.

One issue: It's syntactically and semantically legal right now to omit a `break` statement from each branch. This is obviously guaranteed to be a bad thing as the implied `instanceof` check and cast would only be valid in one branch and not the branches that follow it.

* Add some metadata in the class file to mark a class as algebraic. It's possible that each case of an algebraic type could be assigned an integer identifier and then the `switch` construct above could be compiled to a `tableswitch` instruction as switches generally are already (this is how `enum` types are already handled). Each language implementation currently handles the metadata differently, but they pretty much all store proprietary metadata in an annotation on the classes. The main thing that's needed is a way to state that a class is effectively final except for the fixed set of subclasses defined with it. All of the surveyed language implementations of these types generate a fallback case in their `switch/match/case` expression that raises some sort of `MatchException` to deal with out-of-date bytecode (when, for example, you compile your program against version 2 of a library where a type `T` has three cases, but you end up running into [possibly deserialized] values of `T` from older versions of the library where `T` previously had four cases).

We have so many library implementations and at least four language popular implementations that use these types, so a properly written proposal would be tough to ignore.

**Comment by io7m** [comment]
Additional: There could/should be interactions with these types with the value types due in Java 10. Probably best to consult the Valhalla team!

**Comment by jbgi** [comment]
@io7m this is a nice draft. Note that, however, approachs of ADT based on visitor pattern are somewhat more powerful because a visitor is a form of object algebra (cf. http://ropas.snu.ac.kr/~bruno/papers/ecoop2012.pdf). Eg. they make possible the implementation of lazy constructors or the implementation of different algebraic "view" on a particular datatype.

**Comment by io7m** [comment]
@jbgi: Oh, no doubt there's drastically more that can be done with it, and drastically more representations that it wouldn't be able to express. However, given how conservative the Java designers and maintainers are, and how resistant they're likely to be to anything that requires intrusive additions to the language and VM, I think it will be necessary to push for the smallest set of changes that would cover the most common use cases. In my opinion, that's the small set of additions above that would just cover _case analysis_. It's fairly easy to demonstrate the utility of a pure case analysis system (I've just written a small compiler for a documentation language in Kotlin, for example, where the bulk of the code is almost purely `when` expressions performing case analysis). It's somewhat less easy to demonstrate the benefits of eg. lazy constructors outside of the abstract realm that we as proponents of functional programming inhabit. I would also hope that with the standardized class metadata for expressing `switch/case/variant/algebraic` types, the type implementations would become somewhat compatible between languages (for example, declaring a `sealed` class in Kotlin, and being able to pattern match on that same class in Java). This becomes somewhat less likely the more exotic the types that are allowed (for example, it's possible to seal an interface in Scala, but it's not possible to do the same in Kotlin, so a `sealed interface` in Scala would have no compatible representation in Kotlin).
 
Obviously it'd still be possible to implement tools that generate code using the visitor approach!

**Comment by jbgi** [comment]
@io7m I agree. Just wanted to drop a note that sealed class are not strictly equivalent to the visitor pattern. (another problem of sealed hierarchies, often encountered in scala, is that they expose subtyping relationships, which often ruin type inference).

**Comment by io7m** [comment]
> ruin type inference

Hey, perhaps I _wanted_ `Any → Any → Any`!

**Comment by before** [comment]
@io7m having native support in the language itself would be the better approach.
@elucash Java 10 seems to be far away and if the value types will make it into that release is still open. From my perspective it is worth to have an annotation processor based solution until Java supports it natively.

**Comment by alahijani** [comment]
There is a little technique by which ADTs can be represented quite elegantly in pure Java and sometimes I wish it was better known.

I can showcase it with the quintessential example of binary expressions: 

```java
public abstract class Exp {
    private Exp() { }

    public interface ExpVisitor<R> {
        R lit(int value);
        R add(R left, R right);
    }

    public abstract <R> R accept(ExpVisitor<R> visitor);

    public static final ExpVisitor<Exp> factory = createFactory();
    public static final ExpVisitor<Integer> eval = new ExpVisitor<Integer>() {
        @Override public Integer lit(int value) { return value; }
        @Override public Integer add(Integer left, Integer right) { return left + right; }
    };

    public static void main(String[] args) {
        // 1 + (1 + 1)
        Exp exp = factory.add(factory.lit(1), factory.add(factory.lit(1), factory.lit(1)));
        // result: 3
        System.out.println("result: " + exp.accept(eval));
    }
}
```
Personally I think the most elegant part of this technique is the `factory` object and how it's also a visitor. 

The implementation for `createFactory()` is formulaic and actually can be automated with code generation or reflection, but here I have a hand-coded implementation just to make the above example work.
```java
    private static ExpVisitor<Exp> createFactory() {
        return new ExpVisitor<Exp>() {
            @Override public Exp lit(int value) {
                return new Exp() {
                    @Override public <R> R accept(ExpVisitor<R> visitor) {
                        return visitor.lit(value);
                    }
                };
            }

            @Override public Exp add(Exp left, Exp right) {
                return new Exp() {
                    @Override public <R> R accept(ExpVisitor<R> visitor) {
                        return visitor.add(left.accept(visitor), right.accept(visitor));
                    }
                };
            }
        };
    }
```

**Comment by jbgi** [comment]
@alahijani yes, this "corrected" visitor pattern is what https://github.com/derive4j/derive4j actually use to generate algebraic data types in java.
The factory visitor is quite like an identity function and can indeed be useful as a template for creating modifier/setter functions à la lens/prism (by overriding one/some of the methods)
Examples similar to yours, using derive4j:

- https://github.com/derive4j/derive4j#popular-use-case-domain-specific-languages
- https://gist.github.com/jbgi/2d26c2b30b658ea84b3e458cc8e846aa (derive4j version of https://engineering.vena.io/2016/05/30/ocaml-mimicry-in-java-jocaml/)

**Comment by alahijani** [comment]
@jbgi there is a small but crucial difference in the way self-referencing is handled. Compare this:
```java
public interface ExpVisitor<R> {
    R lit(int value);
    R add(R left, R right);
}
```
against:
```java
public interface ExpVisitor<R> {
    R lit(int value);
    R add(Exp left, Exp right);
}
```
In the former (which I prefer) the `accept` method implements a catamorphism for the [F-Algebra](https://bartoszmilewski.com/2013/06/10/understanding-f-algebras/) with no magical syntax involved.

As a result, my `eval` visitor above doesn't need to be explicit about recursion, because recursive case analysis is handled once and for all inside the factory.

**Comment by jbgi** [comment]
@alahijani right, I read too quickly, sorry. Indeed this encoding this neat for recursively defined data type. I have meant to add support for it in derive4j since I attended https://github.com/LambdaCon/2015/blob/master/Programming%20with%20algebras/slides/Programming%20with%20algebras.pdf

Derive4j does derive catamorphisms so you don't have to be explicit about recursion either, but it is a bit more heavyweight since the recursion is not built-in.

There is however two problems with that encoding in java:

- Java evaluation is strict, so to be as general as in Haskell we rather need:
```java
public interface ExpVisitor<R> {
    R lit(int value);
    R add(Supplier<R> left, Supplier<R> right);
}
```
- stack frames are limited, so this encoding would blow the stack for deep data structures. So we may need a version that works via trampolining.
